File: python_workspace/gap_detection.py
"""
制动间隙点检测模块

此模块提供制动间隙点识别功能，可独立启用或禁用。
制动间隙点是指电机电流发生突变的位置，通常对应于制动系统开始接触的时刻。

主要功能:
1. detect_gap_point: 检测电流序列中的突变点
2. compensate_displacement_gap: 将检测到的间隙点作为位移零点进行补偿
"""

import numpy as np


def moving_average(data, window):
    """移动平均滤波"""
    if window <= 1:
        return data
    kernel = np.ones(window) / window
    return np.convolve(data, kernel, mode='same')


def detect_gap_point(iq_series, threshold_factor=2.0, window=5):
    """
    检测制动间隙点（电流突变点）
    
    参数:
        iq_series: 电流序列
        threshold_factor: 突变检测阈值倍数（相对于初始电流标准差）
        window: 滑动窗口大小
    
    返回:
        gap_index: 间隙点索引（电流第一次突变的位置）
    """
    # 计算电流的绝对值
    iq_abs = np.abs(iq_series)
    
    # 计算初始段的基线（前10%的数据）
    baseline_len = max(int(len(iq_abs) * 0.1), 50)
    baseline = iq_abs[:baseline_len]
    baseline_mean = np.mean(baseline)
    baseline_std = np.std(baseline)
    
    # 如果标准差太小，使用固定阈值
    if baseline_std < 0.1:
        baseline_std = 0.1
    
    # 计算滑动窗口内的电流变化率
    iq_smooth = moving_average(iq_abs, window)
    delta_iq = np.diff(iq_smooth)
    
    # 寻找第一个显著突变点
    threshold = baseline_mean + threshold_factor * baseline_std
    
    # 从baseline之后开始检测
    search_start = baseline_len
    for i in range(search_start, len(iq_smooth) - 1):
        # 检测电流是否超过阈值，且有明显上升趋势
        if iq_smooth[i] > threshold and delta_iq[i-1] > 0:
            # 向前回溯找到实际起跳点
            for j in range(i, max(search_start - 10, 0), -1):
                if iq_smooth[j] <= baseline_mean + 0.5 * baseline_std:
                    return j
            return i
    
    # 如果没有检测到突变点，返回0（无间隙）
    return 0


def compensate_displacement_gap(x_series, iq_series, gap_detection_params=None):
    """
    补偿制动间隙：将电流突变点作为位移零点
    
    参数:
        x_series: 原始位移序列
        iq_series: 电流序列
        gap_detection_params: 间隙检测参数字典
            - threshold_factor: 阈值倍数（默认2.0）
            - window: 滑动窗口大小（默认5）
    
    返回:
        x_compensated: 补偿后的位移
        gap_index: 检测到的间隙点索引
        gap_value: 间隙大小（位移值）
    """
    if gap_detection_params is None:
        gap_detection_params = {'threshold_factor': 2.0, 'window': 5}
    
    # 检测间隙点
    gap_index = detect_gap_point(iq_series, **gap_detection_params)
    
    # 获取间隙大小（起始位移值）
    gap_value = x_series[gap_index] if gap_index > 0 else 0.0
    
    # 补偿位移：将间隙点作为零点
    x_compensated = x_series - gap_value
    
    # 确保位移非负
    x_compensated = np.maximum(x_compensated, 0.0)
    
    return x_compensated, gap_index, gap_value

File: c_workspace/gap_detection.c
/*
 * 制动间隙点检测模块实现
 */

#include "gap_detection.h"

/* 移动平均滤波 - 本地实现 */
static void gap_moving_average(const float *input, float *output, int len, int window) {
    if (window <= 1) {
        for (int i = 0; i < len; i++) {
            output[i] = input[i];
        }
        return;
    }
    
    int half_win = window / 2;
    for (int i = 0; i < len; i++) {
        int start = (i - half_win < 0) ? 0 : i - half_win;
        int end = (i + half_win >= len) ? len - 1 : i + half_win;
        
        float sum = 0.0f;
        int count = 0;
        for (int j = start; j <= end; j++) {
            sum += input[j];
            count++;
        }
        output[i] = sum / count;
    }
}

/* 检测制动间隙点 */
int detect_gap_point(const float *iq_series, int len, float threshold_factor, int window) {
    if (len < 100) return 0;  // 数据太少，无法检测
    
    // 计算电流绝对值
    float *iq_abs = (float *)malloc(len * sizeof(float));
    for (int i = 0; i < len; i++) {
        iq_abs[i] = fabsf(iq_series[i]);
    }
    
    // 计算基线（前10%的数据）
    int baseline_len = len / 10;
    if (baseline_len < 50) baseline_len = 50;
    if (baseline_len > len / 2) baseline_len = len / 2;
    
    float baseline_sum = 0.0f;
    for (int i = 0; i < baseline_len; i++) {
        baseline_sum += iq_abs[i];
    }
    float baseline_mean = baseline_sum / baseline_len;
    
    // 计算基线标准差
    float baseline_var = 0.0f;
    for (int i = 0; i < baseline_len; i++) {
        float diff = iq_abs[i] - baseline_mean;
        baseline_var += diff * diff;
    }
    float baseline_std = sqrtf(baseline_var / baseline_len);
    
    // 最小标准差
    if (baseline_std < 0.1f) {
        baseline_std = 0.1f;
    }
    
    // 应用移动平均滤波
    float *iq_smooth = (float *)malloc(len * sizeof(float));
    gap_moving_average(iq_abs, iq_smooth, len, window);
    
    // 计算阈值
    float threshold = baseline_mean + threshold_factor * baseline_std;
    
    // 从baseline之后开始检测
    int gap_index = 0;
    for (int i = baseline_len; i < len - 1; i++) {
        // 检测电流超过阈值且有上升趋势
        float delta = iq_smooth[i] - iq_smooth[i - 1];
        if (iq_smooth[i] > threshold && delta > 0) {
            // 向前回溯找实际起跳点
            for (int j = i; j >= baseline_len - 10 && j >= 0; j--) {
                if (iq_smooth[j] <= baseline_mean + 0.5f * baseline_std) {
                    gap_index = j;
                    break;
                }
            }
            if (gap_index == 0) gap_index = i;
            break;
        }
    }
    
    free(iq_abs);
    free(iq_smooth);
    
    return gap_index;
}

/* 补偿制动间隙 */
int compensate_displacement_gap(float *x_series, const float *iq_series, int len,
                                float threshold_factor, int window, float *gap_value) {
    // 检测间隙点
    int gap_index = detect_gap_point(iq_series, len, threshold_factor, window);
    
    // 获取间隙大小
    *gap_value = (gap_index > 0) ? x_series[gap_index] : 0.0f;
    
    // 补偿位移：减去间隙值
    for (int i = 0; i < len; i++) {
        x_series[i] -= *gap_value;
        // 确保非负
        if (x_series[i] < 0.0f) {
            x_series[i] = 0.0f;
        }
    }
    
    return gap_index;
}
